### ISSUE TITLE
The old libtorch version is faster than the new one

### ISSUE THREAD

**Issue by xuezu29** [pytorch/pytorch#41177]
My previous program based on  libtorch  1.0.0 Slow down 100ms after the upgrade. And then I find that the same code ran libtorch1.0.0 would be faster than 1.3.0-1.5.1. That's why?

**Comment by t-vi** [comment]
I don't think we can do much here without a reproducing example.

**Comment by glaringlee** [comment]
@xuezu29
Can you share more details? like @t-vi said
Your hardware/software env.
Can example to reproduce. thx

**Comment by xuezu29** [comment]
> @xuezu29
> Can you share more details? like @t-vi said
> Your hardware/software env.
> Can example to reproduce. th x

th x for your reply, The program was written last year based on libtorch 1.0, and recently upgraded libtorch from 1.0 to 1.51. Surprisingly, the inference speed slowed down by 100ms.        GPU:GTX-1660-6G,CPU:i7-4790K, OS:win10, vs2017.

some 1.0.0 code:
'''
std::shared_ptr<torch::jit::script::Module> m_module;
bool Detect::modelInit(const std::string &modelPath)
{
	m_module = torch::jit::load(modelPath);
	assert(m_module != nullptr);
	m_gpu = torch::cuda::is_available();
	if (m_gpu) {
		m_module->to(at::kCUDA);
	}
	m_module->eval();
	return true;
}

torch::Tensor Detect::readImage(const std::string &imgPath)
{
	cv::Mat img, resized_image;
	img = cv::imread(imgPath);
	if (img.empty())
	{
		LOG(ERROR) << "Read detection image failure! file path is: " << imgPath;
		SPDLOG_ERROR("Read detection image failure! file path is: {0}", imgPath);
	}
	if (m_imgHeight < 0 && m_imgWidth < 0) {
		m_imgHeight = img.rows;
		m_imgWidth = img.cols;
		if (m_imgHeight > m_imgWidth)
		{
			LOG(WARNING) << "image width must bigger than image height.image width is: "\
				<< m_imgWidth << ", image height is: " << m_imgHeight << ".";
		}
	}

	float radio = (float)CENTERNET_IMAGE_WIDTH / (float)(m_imgWidth);
	int shortSz = img.rows*radio;

	cv::resize(img, resized_image, cv::Size(CENTERNET_IMAGE_WIDTH, shortSz));
	cv::Mat inputMat = cv::Mat::zeros(CENTERNET_IMAGE_HEIGHT, CENTERNET_IMAGE_WIDTH, CV_8UC3);
	resized_image.copyTo(inputMat.rowRange(0, shortSz));

	cv::Mat img_float;
	inputMat.convertTo(img_float, CV_32F, 1.0 / 255);
	auto m_options = torch::TensorOptions()
		.dtype(torch::kFloat32)
		.is_variable(false);
	auto img_tensor = at::from_blob(img_float.data, { 1,img_float.rows, img_float.cols,3 }, m_options);

	float mean[] = { 0.559,0.559,0.524 };
	float stdData[] = { 0.115,0.114,0.111 };
	auto tmean = at::from_blob(mean, { 1,1,1,3 }, m_options);
	auto tstd = at::from_blob(stdData, { 1,1,1,3 }, m_options);

	img_tensor = (img_tensor - tmean) / tstd;
	img_tensor = img_tensor.permute({ 0,3,1,2 });
	auto img_var = torch::autograd::make_variable(img_tensor, false);
	if (m_gpu) {
		img_var = img_var.to(at::kCUDA);
	}
	return img_var;
}

void Detect::forwardNet(const std::string &imgPath, const bool &isFatglobule)
{
	std::vector<torch::jit::IValue> _inputs;
	torch::Tensor _img_var = readImage(imgPath); // read image.
	_inputs.push_back(_img_var);

	auto _output = m_module->forward(_inputs).toTuple();
	auto res = _output->elements();
	auto hm = res[0].toTensor().sigmoid_(); //score
	auto wh = res[1].toTensor();            //width and height
	auto reg = res[2].toTensor();           //position regression.
        /* some other pose forward process */
      
}
'''
some 1.5.1 code
'''
torch::jit::script::Module m_module;
bool Detect::modelInit(const std::string &modelPath)
{
	try
	{
		m_module = torch::jit::load(modelPath);
	}
	catch (const c10::Error& e)
	{
		//RW_LOG("error loading the model.");
		SPDLOG_ERROR("error loading the model.{}", e.what());
		return false;
	}
	m_gpu = torch::cuda::is_available();
	if (m_gpu) {
		m_module.to(at::kCUDA);
	}
	m_module.eval();
	return true;
}

torch::Tensor Detect::readImage(const std::string &imgPath)
{
	cv::Mat img, resized_image;
	img = cv::imread(imgPath);
	if (img.empty())
	{
		LOG(ERROR) << "Read detection image failure! file path is: " << imgPath;
		SPDLOG_ERROR("Read detection image failure! file path is: {0}", imgPath);
	}
	if (m_imgHeight < 0 && m_imgWidth < 0) {
		m_imgHeight = img.rows;
		m_imgWidth = img.cols;
		if (m_imgHeight > m_imgWidth)
		{
			LOG(WARNING) << "image width must bigger than image height.image width is: "\
				<< m_imgWidth << ", image height is: " << m_imgHeight << ".";
		}
	}

	float radio = (float)CENTERNET_IMAGE_WIDTH / (float)(m_imgWidth);
	int shortSz = img.rows*radio;

	cv::resize(img, resized_image, cv::Size(CENTERNET_IMAGE_WIDTH, shortSz));
	cv::Mat inputMat = cv::Mat::zeros(CENTERNET_IMAGE_HEIGHT, CENTERNET_IMAGE_WIDTH, CV_8UC3);
	resized_image.copyTo(inputMat.rowRange(0, shortSz));

	cv::Mat img_float;
	inputMat.convertTo(img_float, CV_32F, 1.0 / 255);
	auto m_options = torch::TensorOptions()
		.dtype(torch::kFloat32)
		.is_variable(false);
	auto img_tensor = at::from_blob(img_float.data, { 1,img_float.rows, img_float.cols,3 }, m_options);

	float mean[] = { 0.559,0.559,0.524 };
	float stdData[] = { 0.115,0.114,0.111 };
	auto tmean = at::from_blob(mean, { 1,1,1,3 }, m_options);
	auto tstd = at::from_blob(stdData, { 1,1,1,3 }, m_options);

	img_tensor = (img_tensor - tmean) / tstd;
	img_tensor = img_tensor.permute({ 0,3,1,2 });
	auto img_var = torch::autograd::make_variable(img_tensor, false);
	if (m_gpu) {
		img_var = img_var.to(at::kCUDA);
	}
	return img_var;
}

void Detect::forwardNet(const std::string &imgPath, const bool &isFatglobule)
{
	std::vector<torch::jit::IValue> _inputs;
	torch::Tensor _img_var = readImage(imgPath); // read image.
	_inputs.push_back(_img_var);

	auto _output = m_module.forward(_inputs).toTuple();
	auto res = _output->elements();
	auto hm = res[0].toTensor().sigmoid_(); //score
	auto wh = res[1].toTensor();            //width and height
	auto reg = res[2].toTensor();           //position regression.
        /* some other pose forward process */
      
}
'''
The difference is just the loading function torch::jit::load().

**Comment by xuezu29** [comment]
@glaringlee I've sorted out my code, and i have test, which can find it here:[https://github.com/xuezu29/Libtorchtest.git](url)

**Comment by xuezu29** [comment]
I still haven't found a solution.  I want to use the latest version.  Could you help me？thanks a lot! @glaringlee

**Comment by glaringlee** [comment]
@xuezu29 
Sry to reply late. Missed your previous message.
There are many changes from 1.0 to 1.3-1.5, jit, autograd, etc.
Can you put some code to test some specific parts of your code? including:
jit::load(), from_blob(), and readImage()
Compare with the old code, so I can include the right people to investigate.

**Comment by xuezu29** [comment]
> @xuezu29
> Sry to reply late. Missed your previous message.
> There are many changes from 1.0 to 1.3-1.5, jit, autograd, etc.
> Can you put some code to test some specific parts of your code? including:
> jit::load(), from_blob(), and readImage()
> Compare with the old code, so I can include the right people to investigate.

thanks a lot! All code ,weight files and image data can find here: 
https://github.com/xuezu29/Libtorchtest

**Comment by glaringlee** [comment]
@xuezu29 
Can you test the specific parts that I listed above? jit::load(), from_blob() and readimage()?
I don't have a windows machine on hand, I can try find a windows machine and setup the env and run your code to reproduce the result, but it will be much slower. If you can test these parts and let me know which part is slow, I can find the right person to look into this. Let me know whether you can help testing these specific functions, thx.

**Comment by xuezu29** [comment]
@glaringlee  sorry,I didn't catch your meaning. I have test:
**libtorch1.0.0:**
jit::load(): 1193.32ms
from_blob(): 0.1328ms
readimage()：64.06ms
forward(): 161.881ms
total: 227.802ms


**libtorch 1.5.1:**
jit::load(): 417.05ms
from_blob(): 0.2371ms
readimage()：51.88ms
forward(): 287.18ms
total: 339.813ms

Total time does not include model loading (jit::load()).

**Comment by glaringlee** [comment]
@suo 
I saw the forward() is called from a jit Module, not sure whether this is related to jit changes since version 1.3, can you advise? thx.

**Comment by t-vi** [comment]
One thing is not to measure the first run.

**Comment by xuezu29** [comment]
Yeah,  time was the fifth picture.  @t-vi @glaringlee

**Comment by QWERDF007** [comment]
I use libtorch v1.6 and v1.2 to test deeplabv3+, the result of v1.2 is faster than v1.6.

**Comment by glaringlee** [comment]
@QWERDF007 
Can you paste some benchmarks and code pieces?

**Comment by QWERDF007** [comment]
@glaringlee  ![image](https://user-images.githubusercontent.com/46773878/89380232-c3fd2680-d729-11ea-8af7-a39488cf4b2d.png)

**Comment by QWERDF007** [comment]
And I set environment variable CUDA_LAUNCH_BLOCKING=1.

**Comment by QWERDF007** [comment]
ubuntu16.04 server,  i7-8700K + GTX 1080 Ti

**Comment by glaringlee** [comment]
@QWERDF007 
What is the type of `_module`? is that torch::jit::script::Module?

**Comment by QWERDF007** [comment]
yes.  
torch::Device *_device;
_device = new torch::Device(torch::kCUDA, device_id);
torch::jit::script::Module _module;
_module = torch::jit::load(modelfile, *_device);

**Comment by glaringlee** [comment]
This probably related to jit as well. I will add jit label to let jit team take a look at as well. @suo

**Comment by eellison** [comment]
@glaringlee do you mind rebenchmarking on 1.6? We made some changes to the default execution of JIT models in 1.5 that was reverted in the recent release

**Comment by glaringlee** [comment]
@eellison I think the benchmark from @QWERDF007 is generated from 1.6 already.

@xuezu29 Can you run your benchmark on pytorch 1.6 as @eellison requested?

**Comment by QWERDF007** [comment]
The same code run on 8700K+2070 and Debian10:

**Comment by xuezu29** [comment]
> @eellison I think the benchmark from @QWERDF007 is generated from 1.6 already.
> 
> @xuezu29 Can you run your benchmark on pytorch 1.6 as @eellison requested?

The speed of the 1.6.0 is the same as that of the 1.5.1.

**Comment by gemfield** [comment]
You may have a look at https://zhuanlan.zhihu.com/p/363319763 and https://github.com/DeepVAC/libdeepvac/issues/30

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
